### ISSUE TITLE
Psych syntax error changed in JRuby 9.2.13.0

### ISSUE THREAD

**Issue by koic** [jruby/jruby#6365]
**Environment Information**
  
Provide at least:
* % jruby -v
jruby 9.2.13.0 (2.5.7) 2020-08-03 9a89c94bcc Java HotSpot(TM) 64-Bit Server VM 25.5-b02 on 1.8.0_05-b13 +jit [darwin-x86_64]
* Darwin atelier.local 19.6.0 Darwin Kernel Version 19.6.0: Thu Jun 18 20:49:00 PDT 2020; root:xnu-6153.141.1~1/RELEASE_X86_64 x86_64

**Expected Behavior**

This is behavior of JRuby 9.2.12.0 (and MRI 2.5.8).

```console
% ruby -v
jruby 9.2.12.0 (2.5.7) 2020-07-01 db01a49ba6 Java HotSpot(TM) 64-Bit Server VM 25.5-b02 on 1.8.0_05-b13 +jit [darwin-x86_64]
% ruby -ryaml -e "YAML.load('Exclude: **/*_old.rb')"
Psych::SyntaxError: (<unknown>): expected alphabetic or numeric character, but found *(42) while scanning an alias at line 1 column 11
         parse at org/jruby/ext/psych/PsychParser.java:246
  parse_stream at /Users/koic/.rbenv/versions/jruby-9.2.12.0/lib/ruby/stdlib/psych.rb:456
         parse at /Users/koic/.rbenv/versions/jruby-9.2.12.0/lib/ruby/stdlib/psych.rb:390
          load at /Users/koic/.rbenv/versions/jruby-9.2.12.0/lib/ruby/stdlib/psych.rb:277
        <main> at -e:1
```

```console
% ruby -v
ruby 2.5.8p224 (2020-03-31 revision 67882) [x86_64-darwin17]
% ruby -ryaml -e "YAML.load('Exclude: **/*_old.rb')"
Traceback (most recent call last):
        4: from -e:1:in `<main>'
        3: from /Users/koic/.rbenv/versions/2.5.8/lib/ruby/2.5.0/psych.rb:263:in `load'
        2: from /Users/koic/.rbenv/versions/2.5.8/lib/ruby/2.5.0/psych.rb:350:in `parse'
        1: from /Users/koic/.rbenv/versions/2.5.8/lib/ruby/2.5.0/psych.rb:402:in `parse_stream'
/Users/koic/.rbenv/versions/2.5.8/lib/ruby/2.5.0/psych.rb:402:in `parse': (<unknown>): did not find expected alphabetic or numeric character while scanning an alias at line 1 column 10 (Psych::SyntaxError)
```

Both will raise `Psych::SyntaxError`.

**Actual Behavior**

This is behavior of JRuby 9.2.13.0.

```console
% ruby -v
jruby 9.2.13.0 (2.5.7) 2020-08-03 9a89c94bcc Java HotSpot(TM) 64-Bit Server VM 25.5-b02 on 1.8.0_05-b13 +jit [darwin-x86_64]
% ruby -ryaml -e "YAML.load('Exclude: **/*_old.rb')"
Psych::BadAlias: Unknown alias: */*_old.rb
     visit_Psych_Nodes_Alias at /Users/koic/.rbenv/versions/jruby-9.2.13.0/lib/ruby/stdlib/psych/visitors/to_ruby.rb:325
                       fetch at org/jruby/RubyHash.java:1259
     visit_Psych_Nodes_Alias at /Users/koic/.rbenv/versions/jruby-9.2.13.0/lib/ruby/stdlib/psych/visitors/to_ruby.rb:325
                       visit at /Users/koic/.rbenv/versions/jruby-9.2.13.0/lib/ruby/stdlib/psych/visitors/visitor.rb:16
                      accept at /Users/koic/.rbenv/versions/jruby-9.2.13.0/lib/ruby/stdlib/psych/visitors/visitor.rb:6
                      accept at /Users/koic/.rbenv/versions/jruby-9.2.13.0/lib/ruby/stdlib/psych/visitors/to_ruby.rb:34
                 revive_hash at /Users/koic/.rbenv/versions/jruby-9.2.13.0/lib/ruby/stdlib/psych/visitors/to_ruby.rb:343
                  each_slice at org/jruby/RubyArray.java:1851
                 revive_hash at /Users/koic/.rbenv/versions/jruby-9.2.13.0/lib/ruby/stdlib/psych/visitors/to_ruby.rb:341
   visit_Psych_Nodes_Mapping at /Users/koic/.rbenv/versions/jruby-9.2.13.0/lib/ruby/stdlib/psych/visitors/to_ruby.rb:166
                       visit at /Users/koic/.rbenv/versions/jruby-9.2.13.0/lib/ruby/stdlib/psych/visitors/visitor.rb:16
                      accept at /Users/koic/.rbenv/versions/jruby-9.2.13.0/lib/ruby/stdlib/psych/visitors/visitor.rb:6
                      accept at /Users/koic/.rbenv/versions/jruby-9.2.13.0/lib/ruby/stdlib/psych/visitors/to_ruby.rb:34
  visit_Psych_Nodes_Document at /Users/koic/.rbenv/versions/jruby-9.2.13.0/lib/ruby/stdlib/psych/visitors/to_ruby.rb:317
                       visit at /Users/koic/.rbenv/versions/jruby-9.2.13.0/lib/ruby/stdlib/psych/visitors/visitor.rb:16
                      accept at /Users/koic/.rbenv/versions/jruby-9.2.13.0/lib/ruby/stdlib/psych/visitors/visitor.rb:6
                      accept at /Users/koic/.rbenv/versions/jruby-9.2.13.0/lib/ruby/stdlib/psych/visitors/to_ruby.rb:34
                     to_ruby at /Users/koic/.rbenv/versions/jruby-9.2.13.0/lib/ruby/stdlib/psych/nodes/node.rb:50
                        load at /Users/koic/.rbenv/versions/jruby-9.2.13.0/lib/ruby/stdlib/psych.rb:279
                      <main> at -e:1
```

Unexpectedly, JRuby 9.2.13.0 raises `Psych::BadAlias`.

I encountered this issue in RuboCop's CI build.
https://github.com/rubocop-hq/rubocop/pull/8588

**Comment by headius** [comment]
Hmm that's peculiar. The Psych update did include an updated SnakeYAML library which patched an alias-expansion CVE. It may be that it's rejecting this deeper in SnakeYAML where the libyaml version catches it in the C extension. I will poke around.

**Comment by headius** [comment]
The error in CRuby appears to be coming from the YAML parser itself, in scanner.c here:

https://github.com/ruby/psych/blob/39996192cc74a32ac99833c33a14cc54ab2b2396/ext/psych/yaml/scanner.c#L2362

I ran the bad YAML through a couple YAML checkers, and I managed to see *both* results.

The first one was http://www.yamllint.com/, which says it's "optimized for Ruby" so it's probably using Psych. That one produced the same SyntaxError that appears to originate from libyaml:

```
(<unknown>): did not find expected alphabetic or numeric character while scanning an alias at line 1 column 10
```

The second one was https://yamlchecker.com/ but it's not clear what that is using. It produces the same error as JRuby 9.2.13:

```
unidentified alias "*/*_old.rb" at line 1, column 21:
    Exclude: **/*_old.rb
```

A third checker at https://codebeautify.org/yaml-validator also appears to produce the alias error:

```
Error : Reference "*/*_old.rb" does not exist.
Line : Exclude: **/*_old.rb  undefined
```

I am not clear on what the proper error should be. I believe the two errors come from two separate philosophies:

* The syntax error rejects an alias (indicated by `*`) that is not followed by the expected characters for an alias name.
* The alias error rejects the alias later on when it is determined that there's no alias matching the remaining characters.

libyaml still appears to do the former, but other parsers do the latter.

The YAML spec includes this text about aliases:

  It is an error for an alias node to use an anchor that does not previously occur in the document.

So it seems there's some ambiguity here about whether this peculiar alias should be a syntax error or an alias error.

**Comment by headius** [comment]
Also from the YAML spec... for the "anchor" part of the alias, there is only this restriction:

> Anchor names must not contain the “[”, “]”, “{”, “}” and “,” characters. These characters would cause ambiguity with flow collection structures.

And the name of the anchor can contain any non-space character, if I'm reading right.

I'm starting to lean toward the alias error being more correct.

https://yaml.org/spec/1.2/spec.html#id2786196

@koic Can you show me the build or test that failed in RuboCop? I am not sure how to find it.

**Comment by koic** [comment]
@headius Thank you for the investigating and sorry for the missing information. 
The YAML that became the issue is below:

```yaml
AllCops:
  Exclude:
    - **/*_old.rb
```

And the following is the original test that failed on JRuby 9.2.13.0.
https://github.com/rubocop-hq/rubocop/blob/362df666da2700bec8d16762ecba5af5fc32db9f/spec/rubocop/cli_spec.rb#L1566-L1580

**Comment by headius** [comment]
@koic Thank you!

At this point I'm unsure how to proceed. In order to match the libyaml behavior, we would need to patch SnakeYAML, and I would need a good justification for them to do that. I can't find anything in the spec that would reject this as a syntax error.

**Comment by headius** [comment]
I think the next step is to file a bug with libyaml itself. That seems to be the canonical implementation, and several of the spec maintainers also monitor libyaml issues.

The issue to report would be asking for clarification of what the correct error should be here: a syntax error or an alias error. Point out that SnakeYAML and libyaml differ here.

* If they agree it should be an alias error, there's no change required for JRuby but Psych will want to update its libyaml at some point.
* If they decide the syntax error is correct, then we will need to file an issue with SnakeYAML pointing out this behavioral difference.

@koic If you don't get around to it, I will... eventually.

**Comment by headius** [comment]
I have filed https://github.com/yaml/libyaml/issues/205 to get clarification from the libyaml maintainers.

**Comment by headius** [comment]
The libyaml folks have spoken, and while they admit libyaml is much more restrictive than the spec, they also say the spec is too permissive. The future plan is most likely going to follow [YAML RFC-0003](https://github.com/yaml/yaml-spec/blob/master/rfc/RFC-0003.md) which restricts alias to characters useful in an alias (more than libyaml, less than SnakeYAML).

I have filed an issue with @asomov to evaluate SnakeYAML behavior.

https://bitbucket.org/asomov/snakeyaml/issues/485/alias-names-are-too-permissive-compared-to

**Comment by headius** [comment]
With a prerelease build of SnakeYAML 1.28, the original error appears to be restored:

```
$ ruby -ryaml -e "YAML.load('Exclude: **/*_old.rb')"
Psych::SyntaxError: (<unknown>): unexpected character found *(42) while scanning an alias at line 1 column 11
         parse at org/jruby/ext/psych/PsychParser.java:246
  parse_stream at /Users/headius/projects/jruby/lib/ruby/stdlib/psych.rb:454
         parse at /Users/headius/projects/jruby/lib/ruby/stdlib/psych.rb:388
          load at /Users/headius/projects/jruby/lib/ruby/stdlib/psych.rb:277
        <main> at -e:1
```

I will get a release of Psych out and JRuby updated once SnakeYAML 1.28 is released.

**Comment by koic** [comment]
Great news! Thank you for the sharing the progress. I looking forward to releasing JRuby 9.2.14.0 :-)

**Comment by headius** [comment]
Unfortunately it looks like we just missed the last semi-yearly release of SnakeYAML that @asomov did in early September, before this issue was discovered.

The next release will be in February so it will not be possible to fix this until then. I will punt it through versions until we can get an updated SnakeYAML into a release.

**Comment by headius** [comment]
SnakeYAML 1.28 was released yesterday so we can update psych now.

**Comment by headius** [comment]
With a prerelease psych 3.3.0:

```
$ ruby -v -ryaml -e "YAML.load('Exclude: **/*_old.rb')"
jruby 9.2.15.0-SNAPSHOT (2.5.7) 2021-02-23 9f51cbd5cf OpenJDK 64-Bit Server VM 11.0.4+11 on 11.0.4+11 +jit [darwin-x86_64]
Psych::SyntaxError: (<unknown>): unexpected character found *(42) while scanning an alias at line 1 column 11
         parse at org/jruby/ext/psych/PsychParser.java:246
  parse_stream at /Users/headius/projects/jruby-9.2/lib/ruby/gems/shared/gems/psych-3.3.0-java/lib/psych.rb:456
         parse at /Users/headius/projects/jruby-9.2/lib/ruby/gems/shared/gems/psych-3.3.0-java/lib/psych.rb:390
          load at /Users/headius/projects/jruby-9.2/lib/ruby/gems/shared/gems/psych-3.3.0-java/lib/psych.rb:280
        <main> at -e:1
```

**Comment by headius** [comment]
Will resolve once psych release is out and 9.2 points at it.

**Comment by headius** [comment]
This is likely to slip past 9.2.15.0 since it needs a psych release and we will not have time to get that and also update 9.2. Users will still be able to install the updated psych when it is released, since it is a default gem.

Leaving this open for the moment, but it does not need to hold up the release of 9.2.15.0.

See ruby/psych#477.

**Comment by headius** [comment]
Psych 3.3.1 breaks the test suite, probably because it has not been passing on JRuby for some time. I suspect the JRuby extension has gotten out of sync with the C extension.

https://github.com/ruby/psych/issues/479

I am reverting Psych update for now.

**Comment by asomov** [comment]
@headius unfortunately we do not have enough information to resolve https://bitbucket.org/asomov/snakeyaml/issues/498/clarifying-differences-from-libyaml
If you feel that something should be fixed (improved) in SnakeYAML please create a ticket per deviation.

**Comment by headius** [comment]
This should be fixed by #7600 once we get that into 9.3.11.

**Comment by headius** [comment]
Retargeting 9.4.1.0 because it will be the first release with Psych 5.1 and this fix.
